### ISSUE TITLE
Type-Error because of 'declare module "@vue/runtime-core"'

### ISSUE THREAD

**Issue by delphi-sucks** [quasarframework/quasar#17416]
### What happened?

If one defines a global variable like described in the [vue documentation](https://vuejs.org/guide/typescript/options-api.html#augmenting-global-properties) within a typescript project with `define module "vue" { ... }`, then all quasar components don't get resolved within the template anymore.

This results in `vue-tsc` throwing errors and VSCode, with the official Vue extension, not being able to resolve the component to show hints.

### What did you expect to happen?

The quasar components should still resolve correctly if a user (or another package) defines a global vue variable.

### Reproduction URL

https://github.com/delphi-sucks/quasar-bug-define-module-vue

### How to reproduce?

1. Create a new quasar project with **typescript** and `vue-tsc` with at least version 2.0.29
2. Create a new file with a global Vue variable
3. Run vue-tsc (`vue-tsc -p tsconfig.vue-tsc.json`) or open a vue-file which uses a quasar component and verify that the component doesn't get resolved (`unknown`)

**example-global-var-definition.d.ts**

```ts
declare module 'vue' {
  interface GlobalComponents {
    // ...
  }
}

export default {};
```

### Flavour

Quasar CLI with Vite (@quasar/cli | @quasar/app-vite)

### Areas

TypeScript Support

### Platforms/Browsers

Other

### Quasar info output

```Shell
Operating System - Windows_NT(10.0.22631) - win32/x64
NodeJs - 20.12.2

Global packages
  NPM - 10.8.2
  yarn - 1.22.22
  pnpm - 9.6.0
  bun - Not installed
  @quasar/cli - undefined
  @quasar/icongenie - Not installed
  cordova - Not installed

Important local packages
  quasar - 2.16.7 -- Build high-performance VueJS user interfaces (SPA, PWA, SSR, Mobile and Desktop) in record time
  @quasar/app-vite - 2.0.0-beta.15 -- Quasar Framework App CLI with Vite
  @quasar/extras - 1.16.12 -- Quasar Framework fonts, icons and animations
  eslint-plugin-quasar - Not installed
  vue - 3.4.35 -- The progressive JavaScript framework for building modern web UI.
  vue-router - 4.4.2
  pinia - Not installed
  vuex - Not installed
  vite - 5.3.5 -- Native-ESM powered web dev build tool
  vite-plugin-checker - Not installed
  eslint - 8.57.0 -- An AST-based pattern checker for JavaScript.
  esbuild - 0.23.0 -- An extremely fast JavaScript and CSS bundler and minifier.
  typescript - 5.5.4 -- TypeScript is a language for application scale JavaScript development
  workbox-build - Not installed
  register-service-worker - Not installed
  electron - Not installed
  electron-packager - Not installed
  @electron/packager - Not installed
  electron-builder - Not installed
  @capacitor/core - Not installed
  @capacitor/cli - Not installed
  @capacitor/android - Not installed
  @capacitor/ios - Not installed

Quasar App Extensions
  *None installed*

Networking
  *Not relevant*
```

### Relevant log output

```Shell
> quasar-project@0.0.1 typecheck
> vue-tsc --noEmit -p tsconfig.vue-tsc.json

src/pages/IndexPage.vue:9:38 - error TS7006: Parameter 'event' implicitly has an 'any' type.

9     <q-btn label="Press me" @click="(event) => console.log(event)" />
                                       ~~~~~


Found 1 error in src/pages/IndexPage.vue:9
```

### Additional context

This is a problem which was introduced with the latest vue/vue-tsc/volar changes.

Projects are slowly migrating from `define module "@vue/runtime-core"` to `define module "vue"` like [vue-router](https://github.com/vuejs/router/pull/2295) and having packages with different ways to implement global vue variables _can_ result in unexpected behaviour like quasar components not being resolvable anymore.

**Comment by Wangijun** [comment]
The same problem has occurred, and it has now affected normal development

**Comment by yusufkandemir** [comment]
Thanks for reporting. It didn't work for me even without the `GlobalComponents` custom augment in the project. So, the issue is not specific to users that use custom augmentation.
According to https://github.com/vuejs/language-tools/issues/4170#issuecomment-2025528945, augmenting only one of the packages can't guarantee it will work. So, we should augment all `vue`, `@vue/runtime-core`, and `@vue/runtime-dom`. I will handle this.

**Comment by Wangijun** [comment]
![image](https://github.com/user-attachments/assets/fdb6f37b-dc99-4d84-be3f-fcdfa302441c)

My current situation is that the component can be parsed and there are no errors during development, but after running it, vue tsc reports such an error. All of these writing methods result in vue tsc reporting an error.

**Comment by Wangijun** [comment]
> ![image](https://private-user-images.githubusercontent.com/40266995/355372689-fdb6f37b-dc99-4d84-be3f-fcdfa302441c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5MzMyNzksIm5iZiI6MTcyMjkzMjk3OSwicGF0aCI6Ii80MDI2Njk5NS8zNTUzNzI2ODktZmRiNmYzN2ItZGM5OS00ZDg0LWJlM2YtZmNkZmEzMDI0NDFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA2VDA4MjkzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThjNmY1YzJlZTM2ZTMxNjBjNmZkZDdmZDc2ZDk0ZjdjNTMxM2FhNjI5ODI0MmEyYjk3OGNmZDNhMDgwMzlkMjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0._k7QwQJPQTLAtAocsdV9NnhxQW2DTohWMqE7MWVX2KY)
> 
> 我目前的情况是，组件可以解析，且开发中没有错误，但是运行后，vue-tsc，报这样的错误，所有这种写法，vue-tsc都报错。

My current situation is that the component can be parsed and there are no errors during development, but after running it, vue tsc reports such an error. All of these writing methods result in vue tsc reporting an error.

**Comment by delphi-sucks** [comment]
> ![image](https://private-user-images.githubusercontent.com/40266995/355372689-fdb6f37b-dc99-4d84-be3f-fcdfa302441c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5MzM3OTEsIm5iZiI6MTcyMjkzMzQ5MSwicGF0aCI6Ii80MDI2Njk5NS8zNTUzNzI2ODktZmRiNmYzN2ItZGM5OS00ZDg0LWJlM2YtZmNkZmEzMDI0NDFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA2VDA4MzgxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZkMTA5MDNkYjBkOTcxZTUxYjBiZDE0YTg4ODExNDVmY2FhMjBhNWI5MzFmYWViYWVkZmJhODJkNDc4NWVhM2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.T87c6olh0nOV63-e4z0gpnbPo7NGDUpJb6bdO_cmEuc)
> 
> My current situation is that the component can be parsed and there are no errors during development, but after running it, vue tsc reports such an error. All of these writing methods result in vue tsc reporting an error.

Are you possibly using an older version of the VSCode Extension for vue "Vue - Official" (`vue.volar`)? Mine stopped automatically updating at some point and with the older versions the component still gets resolved correctly. I guess if you update to the latest version (v2.0.28), you will also see the problem in VSCode.

**Comment by Wangijun** [comment]
> > ![image](https://private-user-images.githubusercontent.com/40266995/355372689-fdb6f37b-dc99-4d84-be3f-fcdfa302441c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5MzM3OTEsIm5iZiI6MTcyMjkzMzQ5MSwicGF0aCI6Ii80MDI2Njk5NS8zNTUzNzI2ODktZmRiNmYzN2ItZGM5OS00ZDg0LWJlM2YtZmNkZmEzMDI0NDFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODA2VDA4MzgxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZkMTA5MDNkYjBkOTcxZTUxYjBiZDE0YTg4ODExNDVmY2FhMjBhNWI5MzFmYWViYWVkZmJhODJkNDc4NWVhM2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.T87c6olh0nOV63-e4z0gpnbPo7NGDUpJb6bdO_cmEuc)
> > My current situation is that the component can be parsed and there are no errors during development, but after running it, vue tsc reports such an error. All of these writing methods result in vue tsc reporting an error.
> 
> Are you possibly using an older version of the VSCode Extension for vue "Vue - Official" (`vue.volar`)? Mine stopped automatically updating at some point and with the older versions the component still gets resolved correctly. I guess if you update to the latest version (v2.0.28), you will also see the problem in VSCode.


I am currently in version v2.0.28, and only the vue tsc error appears in the console after executing quasar dev, but there is no error prompt in the editor,

**Comment by yusufkandemir** [comment]
The fix will be included in the next `quasar` package release.

**Comment by Logima** [comment]
This still seems to be broken when a project augments ComponentCustomProperties and uses custom augmentation from another project, e.g. vue-router.

[Here](https://github.com/Logima/quasar-component-custom-properties) is a repro.

```
$ pnpm vue-tsc --noEmit -p tsconfig.vue-tsc.json
src/pages/IndexPage.vue:58:19 - error TS2339: Property '$route' does not exist on type 'CreateComponentPublicInstance<{}, {}, { todos: Todo[]; meta: Meta; }, {}, { getRoutePath(): string; }, ComponentOptionsMixin, ComponentOptionsMixin, ... 12 more ..., {}>'.

58       return this.$route.fullPath;
                     ~~~~~~


Found 1 error in src/pages/IndexPage.vue:58
```

Reference to `this.$route` fails unless [Quasar's ComponentCustomProperties](https://github.com/quasarframework/quasar/blob/dev/ui/build/build.types.js#L623) augmentation is changed from `declare module \'@vue/runtime-core\'` to `declare module \'vue\'`.

Could @yusufkandemir take another look at this please?

**Comment by yusufkandemir** [comment]
@Logima you are right, we've overlooked that part. For simplicity purposes, I will just reopen this issue instead of creating a new one.

**Comment by Zeph33** [comment]
in quasar/dist/types/index.d.ts
```javascript
declare module "@vue/runtime-core" {
   interface ComponentCustomProperties {
     $q: QVueGlobals;
```
Fix buid.types.js
[quasar/ui/build/build.types.js](https://github.com/quasarframework/quasar/blob/dev/ui/build/build.types.js)
replace line 622
```
  // Extend Vue instance with injections
  writeLine(contents, 'declare module \'@vue/runtime-core\' {')
  writeLine(contents, 'interface ComponentCustomProperties {', 1)
```
by
```
  // Extend Vue instance with injections
  writeLine(contents, 'declare module \'vue\' {')
  writeLine(contents, 'interface ComponentCustomProperties {', 1)
```

**Comment by skromm** [comment]
Do you have any news about a new release of Quasar solving this bug? Can you propose a workaround?

**Comment by rstoenescu** [comment]
Fix will be available in Quasar v2.17+
